kong/db/dao/plugins/go.lua

```diff
@@ -0,0 +1,486 @@
+        while sent_bytes < #s do
+          local rc = C.write(self.fd, p+sent_bytes, #s-sent_bytes)
+          if rc < 0 then
+            local errno = ffi.errno()
+            return nil, "error writing to socket: " .. ffi.string(C.strerror(errno))
+          end
+          sent_bytes = sent_bytes + rc
+        end
+        return sent_bytes
+      end,
+
+      receiveany = function (self, n)
+        local buf = ffi.new('uint8_t[?]', n)
+        if buf == nil then
+          return nil, "can't allocate buffer."
+        end
+
+        local rc = C.read(self.fd, buf, n)
+        if rc < 0 then
+          return nil, "error reading"
+        end
+        return ffi.string(buf, rc)
+      end,
+
+      setkeepalive = function (self)
+        if self.fd ~= -1 then
+          C.close(self.fd)
+          self.fd = -1
+        end
+      end,
+    },
+
+    __gc = function (self)
+      self:setkeepalive()
+    end,
+  })
+
+  local function un_addr(path)
+    local conaddr = ffi.new('struct sockaddr_un')
+    if conaddr == nil then
+      return nil, "can't allocate sockaddr_un"
+    end
+
+    conaddr.sun_family = C.AF_UNIX
+
+    local len = #path + 1
+    if len > 107 then
+      len = 107
+    end
+    ffi.copy(conaddr.sun_path, path, len)
+    return conaddr
+  end
+
+  local too_early = {
+    init = true,
+    init_worker = true,
+    set = true,
+    header_filter = true,
+    body_filter = true,
+--     log = true,
+  }
+
+  function get_connection(socket_path)
+    if not too_early[ngx.get_phase()] then
+--       ngx.log(ngx.DEBUG, "not early anymore, use a cosocket!")
+      return ngx.socket.connect("unix:" .. socket_path)
+    end
+
+    local fd = C.socket(C.AF_UNIX, C.SOCK_STREAM, 0)
+    if fd < 0 then
+      return nil, "can't create socket"
+    end
+
+    local conaddr = un_addr(socket_path)
+    local res = C.connect(fd, conaddr, ffi.sizeof(conaddr))
+    if res < 0 then
+      C.close(fd)
+      local errno = ffi.errno()
+      return nil, "connect failure: " .. ffi.string(C.strerror(errno))
+    end
+
+    return ffi_sock(fd)
+  end
+end
+go.get_connection = get_connection
+
+
+-- This is the MessagePack-RPC implementation
+local rpc_call
+local set_plugin_dir
+do
+  local msg_id = 0
+
+  local notifications = {}
+
+  local current_plugin_dir
+
+  do
+    local pluginserver_pid
+    function notifications.serverPid(n)
+      n = tonumber(n)
+      if pluginserver_pid and n ~= pluginserver_pid then
+        reset_instances()
+        current_plugin_dir = nil
+      end
+
+      pluginserver_pid = n
+    end
+  end
+
+  function rpc_call(method, ...)
+    msg_id = msg_id + 1
+    local my_msg_id = msg_id
+
+    local c = assert(get_connection(kong.configuration.pluginserver_socket))
+    local bytes, err = c:send(mp_pack({0, my_msg_id, method, {...}}))
+    if not bytes then
+      c:setkeepalive()
+      return nil, err
+    end
+
+    local reader = mp_unpacker(function ()
+      return c:receiveany(4096)
```

Comment:
it's a chunck-reading.  will be called until it gets a complete object (which in mp-rpc is a 3- or 4-element array)  any "extra" is retained for the next round; so it could (in theory) get a bit complex if we multiplex the connections, but we're not doing that here.